### PR TITLE
#25583 : D3 color 1.2.0

### PR DESCRIPTION
**Pull request by Ledragon**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: [d3-color relases](https://github.com/d3/d3-color/releases)
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by Ledragon** [comment]
cc @tomwanzek @gustavderdrache @denisname

File: types/d3-color/index.d.ts [hunk 1]
```diff
@@ -27,6 +28,11 @@ export interface ColorCommonInstance {
     brighter(k?: number): this;
     darker(k?: number): this;
     rgb(): RGBColor;
+    /**
+     * Returns a hexadecimal string representing this color.
+     * If this color is not displayable, a suitable displayable color is returned instead. For example, RGB channel values greater than 255 are clamped to 255.
+     */
+   hex(): string;
```
Comment (denisname) [inline types/d3-color/index.d.ts]: Hex must be indentation by two.

File: types/d3-color/index.d.ts [hunk 2]
```diff
@@ -27,6 +28,11 @@ export interface ColorCommonInstance {
...
 }
 
 export interface Color {
```
Comment (denisname) [inline types/d3-color/index.d.ts]: This is the `Color` interface you should add `hex()`.

File: types/d3-color/index.d.ts [hunk 3]
```diff
@@ -82,6 +101,11 @@ export interface LabColor extends Color {
     brighter(k?: number): this;
     darker(k?: number): this;
     rgb(): RGBColor;
+    /**
+     * Returns a hexadecimal string representing this color.
+     * If this color is not displayable, a suitable displayable color is returned instead. For example, RGB channel values greater than 255 are clamped to 255.
+     */
+    hex(): string;
```
Comment (denisname) [inline types/d3-color/index.d.ts]: To remove, inherited from `Color`.

File: types/d3-color/index.d.ts [hunk 4]
```diff
@@ -99,6 +132,11 @@ export interface HCLColor extends Color {
...
```
Comment (denisname) [inline types/d3-color/index.d.ts]: To remove, inherited from `Color`.

File: types/d3-color/index.d.ts [hunk 5]
```diff
@@ -116,6 +160,11 @@ export interface CubehelixColor extends Color {
...
```
Comment (denisname) [inline types/d3-color/index.d.ts]: To remove, inherited from `Color`.

File: types/d3-hsv/index.d.ts [hunk 6]
```diff
@@ -68,6 +68,12 @@ export interface HSVColor extends Color {
      * Returns the RGB equivalent of this color.
      */
     rgb(): RGBColor;
+
+    /**
+     * Returns a hexadecimal string representing this color.
+     * If this color is not displayable, a suitable displayable color is returned instead. For example, RGB channel values greater than 255 are clamped to 255.
+     */
+    hex(): string;
```
Comment (denisname) [inline types/d3-hsv/index.d.ts]: To remove, inherited from `Color`.

File: types/d3-color/index.d.ts [hunk 7]
```diff
@@ -91,6 +115,15 @@ export interface LabColorFactory extends Function {
     readonly prototype: LabColor;
 }
 
+/**
+ * Constructs a new Lab color with the specified l value and a = b = 0.
+ */
+// export interface GrayColorFactory extends Function {
+//     (l: number, opacity?: number): LabColor;
+// }
```
Comment (denisname) [inline types/d3-color/index.d.ts]: Avoid commented code.

File: types/d3-color/index.d.ts [hunk 8]
```diff
@@ -65,6 +79,11 @@ export interface HSLColor extends Color {
     brighter(k?: number): this;
     darker(k?: number): this;
     rgb(): RGBColor;
+    /**
+     * Returns a hexadecimal string representing this color.
+     * If this color is not displayable, a suitable displayable color is returned instead. For example, RGB channel values greater than 255 are clamped to 255.
+     */
+    hex(): string;
```
Comment (denisname) [inline types/d3-color/index.d.ts]: To remove, inherited from `Color`.

File: types/d3-color/index.d.ts [hunk 9]
```diff
@@ -47,7 +53,15 @@ export interface RGBColor extends Color {
...
     darker(k?: number): this;
+    /**
+     * Returns the RGB equivalent of this color.
+     */
     rgb(): this;
+    /**
+     * Returns a hexadecimal string representing this color.
+     * If this color is not displayable, a suitable displayable color is returned instead. For example, RGB channel values greater than 255 are clamped to 255.
+     */
+    hex(): string;
```
Comment (denisname) [inline types/d3-color/index.d.ts]: To remove, inherited from `Color`.

**Review by denisname — COMMENTED** [review]
The `hex` method is still not in the `Color` interface...

File: types/d3-color/index.d.ts (hunk 7 above)
Comment (Ledragon) [reply types/d3-color/index.d.ts]: Actually left on purpose. I used a type as I got an error using the interface. But do we want a type, or add an exception rule to allow for the interface?

File: types/d3-color/index.d.ts (hunk 2 above)
Comment (Ledragon) [reply types/d3-color/index.d.ts]: Shouldn't  `brighter`, `darker`, and `rgb` be moved as well?

Comment (tomwanzek) [reply types/d3-color/index.d.ts]: Only `hex()` would be required to me moved to the abstract base class `Color`. See [Color define statement in source](https://github.com/d3/d3-color/blob/d5258a56a0699ddffd751ff79206cc1399ee7f89/src/color.js#L171). `brighter` and `darker` are actually implemented on the different color space classes.

**Comment by tomwanzek** [comment]
With the recommended change to moving `hex`, this PR is in principle good to go. The only release management question is whether we really want to skip releasing a *Minor 1.1* and jump ahead to *1.2*? We have not done that in the past, even for short-lived consecutive D3 module releases.

**Comment by Ledragon** [comment]
I don't think it matters to have skipped a version. Nobody complained about missing methods using 1.1, and people who will use it from now on will most likely use 1.2, so I guess that's ok.

As for moving things at the interface level, I understand it is not reflecting the code for Color. I however think it could be easier to have a common interface for all Color's extension, that might not be exposed as is, for maintenance's sake. But this is just an idea

File: types/d3-color/index.d.ts [hunk 10]
```diff
@@ -108,6 +132,12 @@ export interface HCLColorFactory extends Function {
     readonly prototype: HCLColor;
 }
 
+export interface LCHColorFactory extends Function {
```
Comment (rbuckton) [inline types/d3-color/index.d.ts]: There is no need to extend `Function`. If an interface has call signatures it automatically gets `Function` as an apparent type.

File: types/d3-color/index.d.ts [hunk 11]
```diff
@@ -91,6 +105,16 @@ export interface LabColorFactory extends Function {
     readonly prototype: LabColor;
 }
 
+/**
+ * Constructs a new Lab color with the specified l value and a = b = 0.
+ */
+export type GrayColorFactory = (l: number, opacity?: number) => LabColor;
+
+// Using the interface creates an error
+// export interface GrayColorFactory extends Function {
```
Comment (rbuckton) [inline types/d3-color/index.d.ts]: This comment should be removed. If you still want to mark this as an interface, note that it is not necessary to extend `Function`.

**Review by rbuckton — CHANGES_REQUESTED** [review]
@Ledragon, in addition to my comments, can you address the remaining reviewer feedback?

**Comment by Ledragon** [comment]
@rbuckton I think all other comments were adressed in the last PR prior to this, maybe @tomwanzek and @denisname can confirm/have a last glance at this? The last pending question was about the bypass of version 1.1, on which we did not make a conclusion

**Comment by rbuckton** [comment]
@Ledragon, I'm referring to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25586#discussion_r187771963, as it seems @denisname was indicating that the additional version of `hex()` was unnecessary since it was already defined on `Color`.

**Comment by Ledragon** [comment]
@rbuckton the it was indeed fixed in the previous PR by moving `hex()` at interface level

**Review by rbuckton — APPROVED** [review]

**Comment by tomwanzek** [comment]
Thanks @Ledragon

**Pull request merged**